File: neumann-ts/src/services/vector.ts
// SPDX-License-Identifier: MIT
/**
 * Vector service clients for points and collections operations.
 */
import type * as grpc from '@grpc/grpc-js';
import type {
  PointsServiceClient,
  CollectionsServiceClient,
  Point as ProtoPoint,
  ScoredPoint as ProtoScoredPoint,
  UpsertPointsResponse,
  GetPointsResponse,
  DeletePointsResponse,
  QueryPointsResponse,
  ScrollPointsResponse,
  CreateCollectionResponse,
  GetCollectionResponse,
  DeleteCollectionResponse,
  ListCollectionsResponse,
} from '../generated/vector.js';
import {
  ConnectionError,
  NotFoundError,
  InternalError,
  InvalidArgumentError,
} from '../types/errors.js';
import type { NeumannError } from '../types/errors.js';

/**
 * A vector point with payload.
 */
export interface VectorPoint {
  /** Unique point identifier. */
  id: string;
  /** Dense vector embedding. */
  vector: number[];
  /** Optional payload data. */
  payload?: Record<string, unknown> | undefined;
}

/**
 * A point with similarity score from a query.
 */
export interface ScoredVectorPoint {
  /** Unique point identifier. */
  id: string;
  /** Similarity score. */
  score: number;
  /** Optional payload data. */
  payload?: Record<string, unknown> | undefined;
  /** Optional vector (if requested). */
  vector?: number[] | undefined;
}

/**
 * Options for upserting points.
 */
export interface UpsertOptions {
  /** Wait for indexing to complete. */
  wait?: boolean;
}

/**
 * Options for getting points.
 */
export interface GetPointsOptions {
  /** Include payload in response. */
  withPayload?: boolean;
  /** Include vector in response. */
  withVector?: boolean;
}

/**
 * Options for querying similar points.
 */
export interface QueryOptions {
  /** Maximum number of results. */
  limit?: number;
  /** Number of results to skip. */
  offset?: number;
  /** Minimum similarity score threshold. */
  scoreThreshold?: number;
  /** Include payload in response. */
  withPayload?: boolean;
  /** Include vector in response. */
  withVector?: boolean;
}

/**
 * Options for scrolling through points.
 */
export interface ScrollOptions {
  /** Maximum number of points to return. */
  limit?: number | undefined;
  /** Offset point ID for pagination. */
  offsetId?: string | undefined;
  /** Include payload in response. */
  withPayload?: boolean | undefined;
  /** Include vector in response. */
  withVector?: boolean | undefined;
}

/**
 * Scroll result with pagination info.
 */
export interface ScrollResult {
  /** Points in this page. */
  points: VectorPoint[];
  /** Next offset ID for pagination, undefined if no more pages. */
  nextOffset?: string | undefined;
}

/**
 * Collection information.
 */
export interface CollectionInfo {
  /** Collection name. */
  name: string;
  /** Number of points in collection. */
  pointsCount: number;
  /** Vector dimension. */
  dimension: number;
  /** Distance metric. */
  distance: string;
}

/**
 * Distance metric for vector similarity.
 */
export type DistanceMetric = 'cosine' | 'euclidean' | 'dot';

/**
 * Service client for vector points operations.
 */
export class PointsClient {
  private client: PointsServiceClient;
  private metadata: grpc.Metadata;

  constructor(client: PointsServiceClient, metadata: grpc.Metadata) {
    this.client = client;
    this.metadata = metadata;
  }

  /**
   * Upsert points into a collection.
   *
   * @param collection - Target collection name.
   * @param points - Points to upsert.
   * @returns Number of points upserted.
   */
  async upsert(collection: string, points: VectorPoint[]): Promise<number> {
    return new Promise((resolve, reject) => {
      const protoPoints: ProtoPoint[] = points.map((p) => ({
        id: p.id,
        vector: p.vector,
        payload: p.payload ? this.encodePayload(p.payload) : undefined,
      }));

      this.client.Upsert(
        { collection, points: protoPoints },
        this.metadata,
        (err: grpc.ServiceError | null, response: UpsertPointsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.upserted);
        }
      );
    });
  }

  /**
   * Get points by IDs.
   *
   * @param collection - Target collection name.
   * @param ids - Point IDs to retrieve.
   * @param options - Get options.
   * @returns Retrieved points.
   */
  async get(
    collection: string,
    ids: string[],
    options: GetPointsOptions = {}
  ): Promise<VectorPoint[]> {
    return new Promise((resolve, reject) => {
      this.client.Get(
        {
          collection,
          ids,
          withPayload: options.withPayload ?? true,
          withVector: options.withVector ?? false,
        },
        this.metadata,
        (err: grpc.ServiceError | null, response: GetPointsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.points.map((p) => this.convertPoint(p)));
        }
      );
    });
  }

  /**
   * Delete points by IDs.
   *
   * @param collection - Target collection name.
   * @param ids - Point IDs to delete.
   * @returns Number of points deleted.
   */
  async delete(collection: string, ids: string[]): Promise<number> {
    return new Promise((resolve, reject) => {
      this.client.Delete(
        { collection, ids },
        this.metadata,
        (err: grpc.ServiceError | null, response: DeletePointsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.deleted);
        }
      );
    });
  }

  /**
   * Query for similar points.
   *
   * @param collection - Target collection name.
   * @param vector - Query vector.
   * @param options - Query options.
   * @returns Similar points with scores.
   */
  async query(
    collection: string,
    vector: number[],
    options: QueryOptions = {}
  ): Promise<ScoredVectorPoint[]> {
    return new Promise((resolve, reject) => {
      this.client.Query(
        {
          collection,
          vector,
          limit: options.limit ?? 10,
          offset: options.offset ?? 0,
          scoreThreshold: options.scoreThreshold,
          withPayload: options.withPayload ?? true,
          withVector: options.withVector ?? false,
        },
        this.metadata,
        (err: grpc.ServiceError | null, response: QueryPointsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.results.map((p) => this.convertScoredPoint(p)));
        }
      );
    });
  }

  /**
   * Scroll through points in a collection.
   *
   * @param collection - Target collection name.
   * @param options - Scroll options.
   * @returns Scroll result with points and next offset.
   */
  async scroll(
    collection: string,
    options: ScrollOptions = {}
  ): Promise<ScrollResult> {
    return new Promise((resolve, reject) => {
      this.client.Scroll(
        {
          collection,
          offsetId: options.offsetId,
          limit: options.limit ?? 100,
          withPayload: options.withPayload ?? true,
          withVector: options.withVector ?? false,
        },
        this.metadata,
        (err: grpc.ServiceError | null, response: ScrollPointsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve({
            points: response.points.map((p) => this.convertPoint(p)),
            nextOffset: response.nextOffset,
          });
        }
      );
    });
  }

  /**
   * Iterate through all points in a collection.
   *
   * @param collection - Target collection name.
   * @param options - Scroll options (limit per page).
   * @returns Async iterable of points.
   */
  async *scrollAll(
    collection: string,
    options: Omit<ScrollOptions, 'offsetId'> = {}
  ): AsyncIterable<VectorPoint> {
    let offsetId: string | undefined;

    while (true) {
      const result = await this.scroll(collection, { ...options, offsetId });
      for (const point of result.points) {
        yield point;
      }
      if (!result.nextOffset) {
        break;
      }
      offsetId = result.nextOffset;
    }
  }

  private convertPoint(proto: ProtoPoint): VectorPoint {
    return {
      id: proto.id,
      vector: proto.vector,
      payload: proto.payload ? this.decodePayload(proto.payload) : undefined,
    };
  }

  private convertScoredPoint(proto: ProtoScoredPoint): ScoredVectorPoint {
    return {
      id: proto.id,
      score: proto.score,
      payload: proto.payload ? this.decodePayload(proto.payload) : undefined,
      vector: proto.vector && proto.vector.length > 0 ? proto.vector : undefined,
    };
  }

  private encodePayload(
    payload: Record<string, unknown>
  ): Record<string, Uint8Array> {
    const result: Record<string, Uint8Array> = {};
    for (const [key, value] of Object.entries(payload)) {
      result[key] = new TextEncoder().encode(JSON.stringify(value));
    }
    return result;
  }

  private decodePayload(
    payload: Record<string, Uint8Array>
  ): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(payload)) {
      try {
        result[key] = JSON.parse(new TextDecoder().decode(value));
      } catch {
        result[key] = new TextDecoder().decode(value);
      }
    }
    return result;
  }

  private handleError(err: grpc.ServiceError): NeumannError {
    const code = err.code as number;
    const NOT_FOUND = 5;
    const INVALID_ARGUMENT = 3;
    const UNAVAILABLE = 14;

    if (code === NOT_FOUND) {
      return new NotFoundError(err.details || 'Collection not found');
    }
    if (code === INVALID_ARGUMENT) {
      return new InvalidArgumentError(err.details || 'Invalid argument');
    }
    if (code === UNAVAILABLE) {
      return new ConnectionError(err.details || 'Service unavailable');
    }
    return new InternalError(err.details || err.message || 'Internal error');
  }
}

/**
 * Service client for vector collections operations.
 */
export class CollectionsClient {
  private client: CollectionsServiceClient;
  private metadata: grpc.Metadata;

  constructor(client: CollectionsServiceClient, metadata: grpc.Metadata) {
    this.client = client;
    this.metadata = metadata;
  }

  /**
   * Create a new collection.
   *
   * @param name - Collection name.
   * @param dimension - Vector dimension.
   * @param distance - Distance metric.
   * @returns True if collection was created.
   */
  async create(
    name: string,
    dimension: number,
    distance: DistanceMetric = 'cosine'
  ): Promise<boolean> {
    return new Promise((resolve, reject) => {
      this.client.Create(
        { name, dimension, distance },
        this.metadata,
        (err: grpc.ServiceError | null, response: CreateCollectionResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.created);
        }
      );
    });
  }

  /**
   * Get collection information.
   *
   * @param name - Collection name.
   * @returns Collection information.
   */
  async get(name: string): Promise<CollectionInfo> {
    return new Promise((resolve, reject) => {
      this.client.Get(
        { name },
        this.metadata,
        (err: grpc.ServiceError | null, response: GetCollectionResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve({
            name: response.name,
            pointsCount: response.pointsCount,
            dimension: response.dimension,
            distance: response.distance,
          });
        }
      );
    });
  }

  /**
   * Delete a collection.
   *
   * @param name - Collection name.
   * @returns True if collection was deleted.
   */
  async delete(name: string): Promise<boolean> {
    return new Promise((resolve, reject) => {
      this.client.Delete(
        { name },
        this.metadata,
        (err: grpc.ServiceError | null, response: DeleteCollectionResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.deleted);
        }
      );
    });
  }

  /**
   * List all collections.
   *
   * @returns Array of collection names.
   */
  async list(): Promise<string[]> {
    return new Promise((resolve, reject) => {
      this.client.List(
        {},
        this.metadata,
        (err: grpc.ServiceError | null, response: ListCollectionsResponse) => {
          if (err) {
            reject(this.handleError(err));
            return;
          }
          resolve(response.collections);
        }
      );
    });
  }

  /**
   * Check if a collection exists.
   *
   * @param name - Collection name.
   * @returns True if collection exists.
   */
  async exists(name: string): Promise<boolean> {
    try {
      await this.get(name);
      return true;
    } catch (err) {
      if (err instanceof NotFoundError) {
        return false;
      }
      throw err;
    }
  }

  private handleError(err: grpc.ServiceError): NeumannError {
    const code = err.code as number;
    const NOT_FOUND = 5;
    const INVALID_ARGUMENT = 3;
    const ALREADY_EXISTS = 6;
    const UNAVAILABLE = 14;

    if (code === NOT_FOUND) {
      return new NotFoundError(err.details || 'Collection not found');
    }
    if (code === INVALID_ARGUMENT) {
      return new InvalidArgumentError(err.details || 'Invalid argument');
    }
    if (code === ALREADY_EXISTS) {
      return new InvalidArgumentError(err.details || 'Collection already exists');
    }
    if (code === UNAVAILABLE) {
      return new ConnectionError(err.details || 'Service unavailable');
    }
    return new InternalError(err.details || err.message || 'Internal error');
  }
}

File: neumann-py/src/neumann/services/vector.py
# SPDX-License-Identifier: MIT
"""Vector service clients for points and collections operations."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal, cast

from neumann.config import ClientConfig, RetryConfig
from neumann.errors import (
    ConnectionError,
    NotFoundError,
)
from neumann.retry import retry_call

if TYPE_CHECKING:
    from types import TracebackType


DistanceMetric = Literal["cosine", "euclidean", "dot"]


@dataclass
class VectorPoint:
    """A vector point with payload."""

    id: str
    vector: list[float]
    payload: dict[str, Any] | None = None


@dataclass
class ScoredVectorPoint:
    """A point with similarity score from a query."""

    id: str
    score: float
    payload: dict[str, Any] | None = None
    vector: list[float] | None = None


@dataclass
class CollectionInfo:
    """Collection information."""

    name: str
    points_count: int
    dimension: int
    distance: str


@dataclass
class ScrollResult:
    """Scroll result with pagination info."""

    points: list[VectorPoint]
    next_offset: str | None = None


class PointsClient:
    """Service client for vector points operations."""

    def __init__(
        self,
        stub: Any,
        metadata: list[tuple[str, str]] | None = None,
        retry_config: RetryConfig | None = None,
    ) -> None:
        self._stub: Any = stub
        self._metadata: list[tuple[str, str]] = metadata or []
        self._retry_config = retry_config or RetryConfig()

    def upsert(
        self,
        collection: str,
        points: list[VectorPoint],
    ) -> int:
        """Upsert points into a collection.

        Args:
            collection: Target collection name.
            points: Points to upsert.

        Returns:
            Number of points upserted.
        """
        from neumann.proto import vector_pb2

        proto_points = []
        for p in points:
            payload = {}
            if p.payload:
                for k, v in p.payload.items():
                    payload[k] = json.dumps(v).encode()
            proto_points.append(vector_pb2.Point(id=p.id, vector=p.vector, payload=payload))

        request = vector_pb2.UpsertPointsRequest(collection=collection, points=proto_points)

        def do_upsert() -> Any:
            return self._stub.Upsert(request, metadata=self._metadata or None)

        response = retry_call(do_upsert, self._retry_config)
        return cast(int, response.upserted)

    def get(
        self,
        collection: str,
        ids: list[str],
        *,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> list[VectorPoint]:
        """Get points by IDs.

        Args:
            collection: Target collection name.
            ids: Point IDs to retrieve.
            with_payload: Include payload in response.
            with_vector: Include vector in response.

        Returns:
            Retrieved points.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.GetPointsRequest(
            collection=collection,
            ids=ids,
            with_payload=with_payload,
            with_vector=with_vector,
        )

        def do_get() -> Any:
            return self._stub.Get(request, metadata=self._metadata or None)

        response = retry_call(do_get, self._retry_config)
        return [self._convert_point(p) for p in response.points]

    def delete(self, collection: str, ids: list[str]) -> int:
        """Delete points by IDs.

        Args:
            collection: Target collection name.
            ids: Point IDs to delete.

        Returns:
            Number of points deleted.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.DeletePointsRequest(collection=collection, ids=ids)

        def do_delete() -> Any:
            return self._stub.Delete(request, metadata=self._metadata or None)

        response = retry_call(do_delete, self._retry_config)
        return cast(int, response.deleted)

    def query(
        self,
        collection: str,
        vector: list[float],
        *,
        limit: int = 10,
        offset: int = 0,
        score_threshold: float | None = None,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> list[ScoredVectorPoint]:
        """Query for similar points.

        Args:
            collection: Target collection name.
            vector: Query vector.
            limit: Maximum number of results.
            offset: Number of results to skip.
            score_threshold: Minimum similarity score threshold.
            with_payload: Include payload in response.
            with_vector: Include vector in response.

        Returns:
            Similar points with scores.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.QueryPointsRequest(
            collection=collection,
            vector=vector,
            limit=limit,
            offset=offset,
            with_payload=with_payload,
            with_vector=with_vector,
        )
        if score_threshold is not None:
            request.score_threshold = score_threshold

        def do_query() -> Any:
            return self._stub.Query(request, metadata=self._metadata or None)

        response = retry_call(do_query, self._retry_config)
        return [self._convert_scored_point(p) for p in response.results]

    def scroll(
        self,
        collection: str,
        *,
        offset_id: str | None = None,
        limit: int = 100,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> ScrollResult:
        """Scroll through points in a collection.

        Args:
            collection: Target collection name.
            offset_id: Offset point ID for pagination.
            limit: Maximum number of points to return.
            with_payload: Include payload in response.
            with_vector: Include vector in response.

        Returns:
            Scroll result with points and next offset.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.ScrollPointsRequest(
            collection=collection,
            limit=limit,
            with_payload=with_payload,
            with_vector=with_vector,
        )
        if offset_id is not None:
            request.offset_id = offset_id

        def do_scroll() -> Any:
            return self._stub.Scroll(request, metadata=self._metadata or None)

        response = retry_call(do_scroll, self._retry_config)
        points = [self._convert_point(p) for p in response.points]
        next_offset: str | None = response.next_offset if response.next_offset else None
        return ScrollResult(points=points, next_offset=next_offset)

    def _convert_point(self, proto: Any) -> VectorPoint:
        """Convert proto point to VectorPoint."""
        payload: dict[str, Any] | None = None
        if proto.payload:
            payload = {}
            for k, v in proto.payload.items():
                try:
                    payload[k] = json.loads(v.decode())
                except (json.JSONDecodeError, UnicodeDecodeError):
                    payload[k] = v.decode()
        return VectorPoint(id=str(proto.id), vector=list(proto.vector), payload=payload)

    def _convert_scored_point(self, proto: Any) -> ScoredVectorPoint:
        """Convert proto scored point to ScoredVectorPoint."""
        payload: dict[str, Any] | None = None
        if proto.payload:
            payload = {}
            for k, v in proto.payload.items():
                try:
                    payload[k] = json.loads(v.decode())
                except (json.JSONDecodeError, UnicodeDecodeError):
                    payload[k] = v.decode()
        vector: list[float] | None = list(proto.vector) if proto.vector else None
        return ScoredVectorPoint(
            id=str(proto.id), score=float(proto.score), payload=payload, vector=vector
        )


class CollectionsClient:
    """Service client for vector collections operations."""

    def __init__(
        self,
        stub: Any,
        metadata: list[tuple[str, str]] | None = None,
        retry_config: RetryConfig | None = None,
    ) -> None:
        self._stub: Any = stub
        self._metadata: list[tuple[str, str]] = metadata or []
        self._retry_config = retry_config or RetryConfig()

    def create(
        self,
        name: str,
        dimension: int,
        distance: DistanceMetric = "cosine",
    ) -> bool:
        """Create a new collection.

        Args:
            name: Collection name.
            dimension: Vector dimension.
            distance: Distance metric.

        Returns:
            True if collection was created.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.CreateCollectionRequest(
            name=name, dimension=dimension, distance=distance
        )

        def do_create() -> Any:
            return self._stub.Create(request, metadata=self._metadata or None)

        response = retry_call(do_create, self._retry_config)
        return cast(bool, response.created)

    def get(self, name: str) -> CollectionInfo:
        """Get collection information.

        Args:
            name: Collection name.

        Returns:
            Collection information.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.GetCollectionRequest(name=name)

        def do_get() -> Any:
            return self._stub.Get(request, metadata=self._metadata or None)

        response = retry_call(do_get, self._retry_config)
        return CollectionInfo(
            name=str(response.name),
            points_count=int(response.points_count),
            dimension=int(response.dimension),
            distance=str(response.distance),
        )

    def delete(self, name: str) -> bool:
        """Delete a collection.

        Args:
            name: Collection name.

        Returns:
            True if collection was deleted.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.DeleteCollectionRequest(name=name)

        def do_delete() -> Any:
            return self._stub.Delete(request, metadata=self._metadata or None)

        response = retry_call(do_delete, self._retry_config)
        return cast(bool, response.deleted)

    def list(self) -> list[str]:
        """List all collections.

        Returns:
            Array of collection names.
        """
        from neumann.proto import vector_pb2

        request = vector_pb2.ListCollectionsRequest()

        def do_list() -> Any:
            return self._stub.List(request, metadata=self._metadata or None)

        response = retry_call(do_list, self._retry_config)
        return [str(c) for c in response.collections]

    def exists(self, name: str) -> bool:
        """Check if a collection exists.

        Args:
            name: Collection name.

        Returns:
            True if collection exists.
        """
        try:
            self.get(name)
            return True
        except NotFoundError:
            return False


class VectorClient:
    """High-level vector client combining points and collections operations."""

    def __init__(
        self,
        address: str,
        *,
        api_key: str | None = None,
        tls: bool = False,
        config: ClientConfig | None = None,
    ) -> None:
        """Initialize vector client (internal use - use connect())."""
        self._address = address
        self._api_key = api_key
        self._tls = tls
        self._config = config or ClientConfig.default()
        self._channel: Any = None
        self._points_stub: Any = None
        self._collections_stub: Any = None
        self._points: PointsClient | None = None
        self._collections: CollectionsClient | None = None
        self._connected = False

    @classmethod
    def connect(
        cls,
        address: str,
        *,
        api_key: str | None = None,
        tls: bool = False,
        config: ClientConfig | None = None,
    ) -> VectorClient:
        """Connect to a remote Neumann server's vector service.

        Args:
            address: Server address in format "host:port".
            api_key: Optional API key for authentication.
            tls: Whether to use TLS encryption.
            config: Optional client configuration.

        Returns:
            A connected VectorClient.

        Raises:
            ConnectionError: If connection fails.
        """
        client = cls(address, api_key=api_key, tls=tls, config=config)

        try:
            import grpc

            options = [
                ("grpc.keepalive_time_ms", client._config.keepalive.time_ms),
                ("grpc.keepalive_timeout_ms", client._config.keepalive.timeout_ms),
                (
                    "grpc.keepalive_permit_without_calls",
                    1 if client._config.keepalive.permit_without_calls else 0,
                ),
            ]

            if tls:
                credentials = grpc.ssl_channel_credentials()
                client._channel = grpc.secure_channel(address, credentials, options=options)
            else:
                client._channel = grpc.insecure_channel(address, options=options)

            from neumann.proto import vector_pb2_grpc

            client._points_stub = vector_pb2_grpc.PointsServiceStub(client._channel)
            client._collections_stub = vector_pb2_grpc.CollectionsServiceStub(client._channel)

            metadata = []
            if api_key:
                metadata.append(("x-api-key", api_key))

            client._points = PointsClient(client._points_stub, metadata, client._config.retry)
            client._collections = CollectionsClient(
                client._collections_stub, metadata, client._config.retry
            )
            client._connected = True

        except ImportError as e:
            raise ConnectionError("gRPC not available. Install with: pip install grpcio") from e
        except Exception as e:
            raise ConnectionError(f"Failed to connect to {address}: {e}") from e

        return client

    @property
    def is_connected(self) -> bool:
        """Check if client is connected."""
        return self._connected

    @property
    def points(self) -> PointsClient:
        """Get the points service client."""
        if self._points is None:
            raise ConnectionError("Client is not connected")
        return self._points

    @property
    def collections(self) -> CollectionsClient:
        """Get the collections service client."""
        if self._collections is None:
            raise ConnectionError("Client is not connected")
        return self._collections

    def close(self) -> None:
        """Close the client connection."""
        if self._channel is not None:
            self._channel.close()
            self._channel = None
        self._points = None
        self._collections = None
        self._connected = False

    def __enter__(self) -> VectorClient:
        """Context manager entry."""
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        """Context manager exit."""
        self.close()

    # Convenience methods that delegate to service clients

    def create_collection(
        self,
        name: str,
        dimension: int,
        distance: DistanceMetric = "cosine",
    ) -> bool:
        """Create a new collection."""
        return self.collections.create(name, dimension, distance)

    def get_collection(self, name: str) -> CollectionInfo:
        """Get collection information."""
        return self.collections.get(name)

    def delete_collection(self, name: str) -> bool:
        """Delete a collection."""
        return self.collections.delete(name)

    def list_collections(self) -> list[str]:
        """List all collections."""
        return self.collections.list()

    def collection_exists(self, name: str) -> bool:
        """Check if a collection exists."""
        return self.collections.exists(name)

    def upsert_points(self, collection: str, points: list[VectorPoint]) -> int:
        """Upsert points into a collection."""
        return self.points.upsert(collection, points)

    def get_points(
        self,
        collection: str,
        ids: list[str],
        *,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> list[VectorPoint]:
        """Get points by IDs."""
        return self.points.get(collection, ids, with_payload=with_payload, with_vector=with_vector)

    def delete_points(self, collection: str, ids: list[str]) -> int:
        """Delete points by IDs."""
        return self.points.delete(collection, ids)

    def query_points(
        self,
        collection: str,
        vector: list[float],
        *,
        limit: int = 10,
        offset: int = 0,
        score_threshold: float | None = None,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> list[ScoredVectorPoint]:
        """Query for similar points."""
        return self.points.query(
            collection,
            vector,
            limit=limit,
            offset=offset,
            score_threshold=score_threshold,
            with_payload=with_payload,
            with_vector=with_vector,
        )

    def scroll_points(
        self,
        collection: str,
        *,
        offset_id: str | None = None,
        limit: int = 100,
        with_payload: bool = True,
        with_vector: bool = False,
    ) -> ScrollResult:
        """Scroll through points in a collection."""
        return self.points.scroll(
            collection,
            offset_id=offset_id,
            limit=limit,
            with_payload=with_payload,
            with_vector=with_vector,
        )

    def count_points(self, collection: str) -> int:
        """Get number of points in a collection."""
        info = self.collections.get(collection)
        return info.points_count
